task/control-plane-agent/src/mgs_sidecar/ignition.rs

```diff
@@ -0,0 +1,298 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+use drv_ignition_api::{
+    AllLinkEventsIter, AllPortsIter, Ignition, IgnitionError,
+};
+use gateway_messages::ignition::{
+    IgnitionState, LinkEvents, ReceiverStatus, SystemFaults, SystemPowerState,
+    SystemType, Target, TransceiverEvents, TransceiverSelect,
+};
+use gateway_messages::{IgnitionCommand, SpError};
+use heapless::Vec;
+use userlib::UnwrapLite;
+
+userlib::task_slot!(IGNITION, ignition);
+
+pub(super) struct IgnitionController {
+    task: Ignition,
+    // We cache the number of ignition ports the first time we successfully call
+    // it since it never changes (it's the total number of ports, which is baked
+    // into the FPGA image, not the number of present targets, which varies at
+    // runtime).
+    num_ports: Option<u32>,
+}
+
+impl IgnitionController {
+    pub(super) fn new() -> Self {
+        Self {
+            task: Ignition::new(IGNITION.get_task_id()),
+            num_ports: None,
+        }
+    }
+
+    pub(super) fn num_ports(&mut self) -> Result<u32, SpError> {
+        if let Some(n) = self.num_ports {
+            return Ok(n);
+        }
+
+        let n = u32::from(self.task.port_count().map_err(map_ignition_error)?);
```

Comment:
If you get tired of `.map_err(map_ignition_error)`, you _could_ use an extension trait, with a blanket implementation, e.g.
```rust
trait MapIgnitionError {
    type V;
    fn map_ignition_error(self) -> Result<Self::V, SpError>;
}

impl<V> MapIgnitionError for Result<V, IgnitionError> {
    type V = V;
    fn map_ignition_error(self) -> Result<Self::V, SpError> {
        self.map_err(|err| {
            use gateway_messages::ignition::IgnitionError as E;
            let err = match err {
                IgnitionError::FpgaError => E::FpgaError,
                IgnitionError::InvalidPort => E::InvalidPort,
                IgnitionError::InvalidValue => E::InvalidValue,
                IgnitionError::NoTargetPresent => E::NoTargetPresent,
                IgnitionError::RequestInProgress => E::RequestInProgress,
                IgnitionError::RequestDiscarded => E::RequestDiscarded,
                _ => E::Other(err as u32),
            };
            SpError::Ignition(err)
        })
    }
}
```
This would allow you to call `.map_ignition_error()` directly.

I don't feel strongly about it, but wanted to present it as an option to consider!

Reply:
After chatting: yes, it's too cute. Reverted to something more explicit/obvious in b95e90b